src/bin/listbox-dnd.rs

```diff
@@ -0,0 +1,228 @@
+    // drag_begin will be called when we start dragging. drag_begin is where we call rendering the
+    // widget being dragged next to mouse cursor.
+    event_box.connect_drag_begin(drag_begin);
+    // We don't use this, but just to demonstrate that things can be done when the drag ends.
+    event_box.connect_drag_end(drag_end);
+    // See https://developer.gnome.org/gtkmm-tutorial/stable/sec-dnd-signals.html.en
+    // Called when a source is being dragged and the target calls get_data()
+    {
+        let r = widget_being_dragged.clone();
+        event_box.connect_drag_data_get(move |w, ctx, sel, i1, i2| {
+            drag_data_get(&mut *r.borrow_mut(), w, ctx, sel, i1, i2)
+        });
+    }
+    // Set the same box as a drag destination.
+    event_box.drag_dest_set(DestDefaults::ALL, &drag_targets, DragAction::MOVE);
+    // Called when get_data() returns
+    {
+        let r = widget_being_dragged.clone();
+        event_box.connect_drag_data_received(move |target, ctx, x, y, sel, info, time| {
+            drag_data_received(
+                r.borrow().as_ref().unwrap(),
+                target,
+                ctx,
+                x,
+                y,
+                sel,
+                info,
+                time,
+            )
+        });
+    }
+
+    list_box_row
+}
+
+fn drag_begin<W: IsA<Widget>>(w: &W, ctx: &DragContext) {
+    println!("drag_begin({:?})", w);
+
+    // Get containing ListBoxRow and render it next to the cursor while being dragged
+    match w.get_ancestor(ListBoxRow::static_type()) {
+        None => {
+            println!("Can't get ancestor");
+        }
+        Some(ancestor) => {
+            let alloc = ancestor.get_allocation();
+            let surface: cairo::ImageSurface =
+                cairo::ImageSurface::create(cairo::Format::ARgb32, alloc.width, alloc.height)
+                    .unwrap();
+            let surface_: &cairo::Surface = surface.as_ref();
+
+            {
+                ancestor.get_style_context().add_class("drag-icon");
+                let surface_ctx = cairo::Context::new(surface_);
+                ancestor.draw(&surface_ctx);
+                ancestor.get_style_context().remove_class("drag-icon");
+            }
+
+            let (x, y) = w.translate_coordinates(&ancestor, 0, 0).unwrap();
+            surface.set_device_offset(-x as f64, -y as f64);
+            ctx.drag_set_icon_surface(surface_);
+        }
+    }
+}
+
+fn drag_end<W: IsA<Widget>>(w: &W, _ctx: &DragContext) {
+    println!("drag_end({:?})", w);
+}
+
+fn drag_data_get<W: IsA<Widget>>(
+    widget_being_dragged: &mut Option<Widget>,
+    w: &W,
+    _ctx: &DragContext,
+    sel: &SelectionData,
+    _: u32,
+    _: u32,
+) {
+    println!("drag_data_get({:?}", w);
+
+    *widget_being_dragged = Some(w.as_ref().clone());
+    // We have to call this to emit drag_data_received() on the target
+    sel.set(&Atom::intern(""), 32, &[]);
```

Comment:
I believe you can simply store the index of the widget inside the listbox here and then later use `get_row_at_index()` to get the widget back from the listbox